// reactchart/lib/Axes.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var React = require('react');
var Axe = require('./axis/Axe.js');
var imUtils = require('./core/im-utils.js');
var _ = require('underscore');

/*
	{
		abs: [Axe],
		ord: [Axe]
	}
*/

var Axes = function (_React$Component) {
	_inherits(Axes, _React$Component);

	function Axes() {
		_classCallCheck(this, Axes);

		return _possibleConstructorReturn(this, (Axes.__proto__ || Object.getPrototypeOf(Axes)).apply(this, arguments));
	}

	_createClass(Axes, [{
		key: 'shouldComponentUpdate',
		value: function shouldComponentUpdate(props) {
			return !imUtils.isEqual(props.state, this.props.state);
		}
	}, {
		key: 'abscissa',
		value: function abscissa() {
			var css = this.props.state.css;
			return _.map(this.props.state.abs, function (p) {
				return p.show ? React.createElement(Axe, { className: 'xAxis', key: p.key, css: css, state: p }) : null;
			});
		}
	}, {
		key: 'ordinate',
		value: function ordinate() {
			var css = this.props.state.css;
			return _.map(this.props.state.ord, function (p) {
				return p.show ? React.createElement(Axe, { className: 'yAxis', key: p.key, css: css, state: p }) : null;
			});
		}
	}, {
		key: 'render',
		value: function render() {

			return React.createElement(
				'g',
				null,
				this.abscissa(),
				this.ordinate()
			);
		}
	}]);

	return Axes;
}(React.Component);

module.exports = Axes;
